Azure/AzureMonitorForSAPSolutions
# Multi-Provider, Multi-Instance

This PR refactors the existing monitoring framework, so that the payload now supports multiple providers (beyond SAP HANA) for which more than one instance can be added to a SAP Monitor resource. Hence, this change addresses two of the biggest shortcomings and opens our platform for additional monitoring providers and content.

MSFT employees can find a detailed design document (including the HanaRP) counterpart in our OneNote.

Apologies for baking this into one large PR, but breaking up all dependencies in an iterative fashion would have significantly slowed down the roll-out. 

**How to test** (until roll-out via Docker image):
- Deploy a SAP Monitor via Portal/AZ CLI
- Reset SSH password and log onto collector VM
- Stop running docker container
- Set up this Git repo
- Ensure required packages are installed by running:
`AzureMonitorForSAPSolutions/master/sapmon/setup/configureMonitorVM.sh`
- Onboard via:
`python3 sapmon.py onboard --logAnalyticsWorkspaceId=xxx --logAnalyticsSharedKey=xxx --providers='[{"name": "onboard", "type": "saphana", "properties": {"HanaHostname": "10.0.0.6", "HanaDbSqlPort": 30113, "HanaDbUsername": "SYSTEM", "HanaDbPassword": "xxx"}}]' `
- Run monitoring via:
`python3 sapmon.py monitor`

**Known limitations:**
- `sapmon.py provider delete` is not implemented yet
- `sapmon.py provider add` does not throw an error if a provider instance with the same name already exists, but rather updates it
- no additional provider types yet (only saphana); next step is to add [prometheus provider](https://github.com/ne-msft/AzureMonitorForSAPSolutions)
